CRAMv3.tex

```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+
+The overall flowchart is below, with more detailed description in the subsequence sections.
+
+\algnewcommand\algorithmicto{\text{ \textbf{to} }}
+
 \subsection{\textbf{CRAM record}}
 
 Both mapped and unmapped reads start with the following fields. Please note that 
 the data series type refers to the logical data type and the data series name corresponds 
 to the data series encoding map.
 
-\begin{tabular}{|>{\raggedright}p{36pt}|>{\raggedright}p{70pt}|>{\raggedright}p{75pt}|>{\raggedright}p{90pt}|>{\raggedright}p{171pt}|}
+\begin{tabular}{|>{\raggedright}p{70pt}|>{\raggedright}p{75pt}|>{\raggedright}p{90pt}|>{\raggedright}p{171pt}|}
+\hline
+\textbf{Data series type} & \textbf{Data series name} & \textbf{Field} & \textbf{Description}\tabularnewline
 \hline
- & \textbf{Data series type} & \textbf{Data series name} & \textbf{Field} & \textbf{Description}\tabularnewline
+int & BF & CRAM bit flags & see CRAM record bit flags\tabularnewline
 \hline
-1 & int & BF & CRAM bit flags & see CRAM record bit flags\tabularnewline
+int & CF  & compression bit flags & see compression bit flags\tabularnewline
 \hline
-2 & int & CF  & compression bit flags & see compression bit flags\tabularnewline
+- & - & Positional data & See section \ref{subsec:positions}\tabularnewline
 \hline
-3 & int & RI & ref id & reference sequence id, not used for single reference slices, 
-reserved for future multiref slices. \tabularnewline
+- & - & Read names & See section \ref{subsec:readnames}\tabularnewline
 \hline
-4 & int & RL & read length & the length of the read\tabularnewline
+- & - & Mate records & See section \ref{subsec:mate}\tabularnewline
 \hline
-5 & int & AP & alignment start & the alignment start position *1\tabularnewline
+- & - & Auxiliary tags & See section \ref{subsec:tags}\tabularnewline
 \hline
-6 & int & RG & read group & the read group identifier\tabularnewline
+- & - & Sequences & See sections \ref{subsec:mapped}and \ref{subsec:unmapped}\tabularnewline
 \hline
-7 & byte & QS & quality scores & quality scores are stored depending on the value 
-of the `mapped QS included' field\tabularnewline
+\end{tabular}
+
+\subsubsection*{\textbf{BAM bit flags (BF data series)}}
+
+The following flags are duplicated from the SAM and BAM specification, with identical meaning.
+Note however some of these flags can be derived during decode, so may be omitted in the CRAM file and the bits computed based on both reads of a pair-end library residing within the same slice.
+
+\begin{tabular}{|>{\raggedright}p{144pt}|>{\raggedright}p{144pt}|>{\raggedright}p{144pt}|}
+\hline
+\textbf{Bit flag} & \textbf{Comment} & \textbf{Description}\tabularnewline
+\hline
+0x1 & ! 0x40 \&\& ! 0x80 & template having multiple segments in sequencing\tabularnewline
```

Comment:
Could you add more to this comment?  I have a few possible interpretations for the correct behavior so I'd like that to be made explicit here.

* Reading
1. bit `0x1` is strictly ignored, but stored as-is for round-tripping equality.  Always use the `0x40` and `0x80` values as the source of truth.
1. store bit `0x1` as-is.  If it conflicts with `0x40` and `0x80`, use the `0x1` value.  
1. bit `0x1` should be validated against the calculation from the `0x40` and `0x80` values, and the record rejected if these do not match
1. bit `0x1` should be corrected (overwritten) by the calculation from the `0x40` and `0x80` values

* Writing
1. write out the `0x1` bit from the `0x40` and `0x80` values, regardless of what it was previously
1. write out the `0x1` bit as the original value, even if it conflicts with the `0x40` and `0x80` values

Reply:
These came from the SAM specification.  I expect Vadim added them because he wanted the CRAM specification to be self contained, which is reasonable.  It's overly terse and I agree confusing.

I'll review them and see what the SAM spec has to say too.  Do we want to just lift the SAM text verbatim, along with all the bullet points elaborating on them?  I think probably we can find a way to add them to the description part of the text, perhaps with footnotes for further explanatory bits.

Reply:
Looking at my C code, I see I store BF verbatim (this is inefficient perhaps).  During read I start with the stored value and then modify it in some manners.

- Add 0x1 if a mate pair is present in this slice.
- Add 0x8 if mate is in this slice and is unmapped
- Add 0x20 if mate is in this slice and reverse complemented.

In other words it copes with these bits being missing and will repopulate (or even correct them if wrong), but it does store them.  I'm unsure why I did it this way.  Maybe the Java code doesn't (or didn't) store all these flags, or perhaps it's related to earlier CRAM versions.  I recall one of them having some bits missing - CRAMv1 maybe - and always generating them.  This doesn't work well though when the mate-pair is outside this slice.

I'd need to check the Java code to see what that does.

Reply:
Setting 0x1 if it has a mate pair fits the description, but I'm not understanding what that has to do with the 0x40 and 0x80 flags.

Reply:
I'm struggling to understand that myself too actually.  I'll look at the Java code - maybe there is a clue there.  Agreed it isn't clear.

Reply:
I just deleted it. :-)

It makes no sense and seems to have no apparent relationship to how either of our implementations work.  I guess it's something lost in the mists of time.